File: program/program/find-the-smallest-two-elements-in-an-array/FindTheSmallestTwoElementsInAnArray.cpp
#include <iostream>
#include <climits>

void findSmallestTwo(int arr[], int size) {
    if (size < 2) {
        std::cout << "Array size should be at least 2." << std::endl;
        return;
    }

    int firstSmallest = INT_MAX;
    int secondSmallest = INT_MAX;

    for (int i = 0; i < size; ++i) {
        if (arr[i] < firstSmallest) {
            secondSmallest = firstSmallest;
            firstSmallest = arr[i];
        } else if (arr[i] < secondSmallest && arr[i] != firstSmallest) {
            secondSmallest = arr[i];
        }
    }

    if (secondSmallest == INT_MAX) {
        std::cout << "There are not enough distinct elements in the array." << std::endl;
    } else {
        std::cout << "The smallest two elements are " << firstSmallest << " and " << secondSmallest << std::endl;
    }
}

int main() {
    int size;

    std::cout << "Enter the size of the array: ";
    std::cin >> size;

    if (size <= 0) {
        std::cout << "Invalid array size." << std::endl;
        return 1;
    }

    int arr[size];
    std::cout << "Enter the elements of the array: ";
    for (int i = 0; i < size; ++i) {
        std::cin >> arr[i];
    }

    findSmallestTwo(arr, size);

    return 0;
}

File: program/program/find-the-smallest-two-elements-in-an-array/find_the_smallest_two_elements_in_an_array.dart
import 'dart:io';

void findSmallestTwo(List<int> arr) {
  if (arr.length < 2) {
    print("Array should have at least two elements.");
    return;
  }

  int smallest = (arr[0] < arr[1]) ? arr[0] : arr[1];
  int secondSmallest = (arr[0] < arr[1]) ? arr[1] : arr[0];

  for (int i = 2; i < arr.length; i++) {
    if (arr[i] < smallest) {
      secondSmallest = smallest;
      smallest = arr[i];
    } else if (arr[i] < secondSmallest) {
      secondSmallest = arr[i];
    }
  }

  print("Smallest two elements in the array: $smallest $secondSmallest");
}

void main() {
  print("Enter integers for the array separated by space:");
  String input = stdin.readLineSync() ?? '';
  List<int> array = input.split(' ').map((e) => int.tryParse(e) ?? 0).toList();

  findSmallestTwo(array);
}


/*

Command Prompt Execution:
dart find_the_smallest_two_elements_in_an_array.dart
Enter integers for the array separated by space:
10 11 9 1 20 3
Smallest two elements in the array: 1 3

*/
